engines/ic3.cpp

```diff
@@ -0,0 +1,299 @@
+  // got through all checks without failing
+  return true;
+}
+
+std::vector<IC3Formula> IC3::inductive_generalization(size_t i,
+                                                      const IC3Formula & c)
+{
+  assert(!c.is_disjunction());  // expecting a cube
+
+  if (options_.mbic3_indgen_mode != 0) {
+    throw PonoException("Boolean IC3 only supports indgen mode 0 but got "
+                        + std::to_string(options_.mbic3_indgen_mode));
+  }
+
+  UnorderedTermSet keep, core_set;
+  TermVec bool_assump, tmp, new_tmp, removed, lits;
+  lits = c.children;
+
+  if (options_.random_seed_ > 0) {
+    shuffle(
+        lits.begin(), lits.end(), default_random_engine(options_.random_seed_));
+  }
+
+  int iter = 0;
+  bool progress = true;
+  while (iter <= options_.ic3_gen_max_iter_ && lits.size() > 1 && progress) {
+    iter = options_.ic3_gen_max_iter_ > 0 ? iter + 1 : iter;
+    size_t prev_size = lits.size();
+    for (auto a : lits) {
+      // check if we can drop a
+      if (keep.find(a) != keep.end()) {
+        continue;
+      }
+      tmp.clear();
+      for (auto aa : lits) {
+        if (a != aa) {
+          tmp.push_back(aa);
+        }
+      }
+
+      Term tmp_and_term = make_and(tmp);
+      if (!intersects_initial(tmp_and_term)) {
+        assert(solver_context_ == 0);
+        push_solver_context();
+        assert_frame_labels(i - 1);
+        assert_trans_label();
+        solver_->assert_formula(solver_->make_term(Not, tmp_and_term));
+
+        Term l;
+        bool_assump.clear();
+        for (auto t : tmp) {
+          l = label(t);
+          solver_->assert_formula(solver_->make_term(Implies, l, ts_->next(t)));
+          bool_assump.push_back(l);
+        }
+
+        Result r = solver_->check_sat_assuming(bool_assump);
+        assert(!r.is_unknown());
+
+        if (r.is_sat()) {
+          // we cannot drop a
+          pop_solver_context();
+        } else {
+          new_tmp.clear();
+          removed.clear();
+          core_set.clear();
+          // filter using unsatcore
+          solver_->get_unsat_core(core_set);
+          for (size_t j = 0; j < bool_assump.size(); ++j) {
+            if (core_set.find(bool_assump[j]) != core_set.end()) {
+              new_tmp.push_back(tmp[j]);
+            } else {
+              removed.push_back(tmp[j]);
+            }
+          }
+
+          pop_solver_context();
+
+          // keep in mind that you cannot drop a literal if it causes c to
+          // intersect with the initial states
+          size_t size = new_tmp.size();
+          fix_if_intersects_initial(new_tmp, removed);
+          // remember the literals which cannot be dropped
+          for (size_t i = size; i < new_tmp.size(); ++i) {
+            keep.insert(new_tmp[i]);
+          }
+
+          lits = new_tmp;
+          break;  // next iteration
+        }
+      }
+    }
+
+    progress = lits.size() < prev_size;
+  }
+
+  // TODO: would it be more intuitive to start with a clause
```

Comment:
@ahmed-irfan flagging this TODO for you. Do you think it would be a cleaner interface for `inductive_generalization` to take a disjunction instead of a conjunction? E.g. take something that's already aligned with the frame and generalize it.

Reply:
i see most algorithms written in papers use cube as input. So, i would stick to that notation. 
Also, imagine, if we don't do any generalization, then this function would return the negation of the cube.